cpp/src/arrow/util/async_generator_test.cc

```diff
@@ -461,6 +461,68 @@ TEST(TestAsyncUtil, Concatenated) {
   AssertAsyncGeneratorMatch(expected, concat);
 }
 
+TEST(TestAsyncUtil, ConcatenatedVector) {
```

Comment:
Given the locks you have it would be good to have at least one parallel test using a readahead generator.